### PR TITLE
fix: kind does not work for array args in compile time

### LINKED ISSUE #4574: Kind does not work for array args in compile time
```fortran
program main
    integer, parameter :: ar1 = kind([1.0, 2.0, 3.0])
    print *, ar1
end program
```
```console
$ gfortran a.f90 && ./a.out
           4
$ lfortran a.f90
LCOMPILERS_ASSERT failed: /home/harshita/Desktop/lfortran/src/libasr/../libasr/pass/pass_utils.h
function replace_ArrayConstant(), line number 855 at 
ASRUtils::extract_n_dims_from_ttype(result_var_type) == 1
ERROR MESSAGE:
Initialisation using ArrayConstant is supported only for single dimensional arrays, found: 0
Traceback (most recent call last):
  Binary file "/home/harshita/Desktop/lfortran/src/bin/lfortran", in _start()
  File "./csu/../csu/libc-start.c", line 392, in __libc_start_main_impl()
  File "./csu/../sysdeps/nptl/libc_start_call_main.h", line 58, in __libc_start_call_main()
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 2481, in ??
    return main_app(argc, argv);
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 2442, in main_app(int, char**)
    err = compile_to_object_file(arg_file, tmp_o, false,
  File "/home/harshita/Desktop/lfortran/src/bin/lfortran.cpp", line 954, in ??
    res = fe.get_llvm3(*asr, lpm, diagnostics, infile);
  File "/home/harshita/Desktop/lfortran/src/lfortran/fortran_evaluator.cpp", line 363, in LCompilers::FortranEvaluator::get_llvm3(LCompilers::ASR::TranslationUnit_t&, LCompilers::PassManager&, LCompilers::diag::Diagnostics&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    compiler_options, run_fn, infile);
  File "/home/harshita/Desktop/lfortran/src/libasr/codegen/asr_to_llvm.cpp", line 10006, in LCompilers::asr_to_llvm(LCompilers::ASR::TranslationUnit_t&, LCompilers::diag::Diagnostics&, llvm::LLVMContext&, Allocator&, LCompilers::PassManager&, LCompilers::CompilerOptions&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    pass_manager.apply_passes(al, &asr, co.po, diagnostics);
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/pass_manager.h", line 315, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    apply_passes(al, asr, _passes, pass_options, diagnostics);
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/pass_manager.h", line 169, in LCompilers::PassManager::apply_passes(Allocator&, LCompilers::ASR::TranslationUnit_t*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, LCompilers::PassOptions&, LCompilers::diag::Diagnostics&)
    _passes_db[passes[i]](al, *asr, pass_options);
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/init_expr.cpp", line 221, in LCompilers::pass_replace_init_expr(Allocator&, LCompilers::ASR::TranslationUnit_t&, LCompilers::PassOptions const&)
    v.visit_TranslationUnit(unit);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 46880, in LCompilers::ASR::CallReplacerOnExpressionsVisitor<LCompilers::InitExprVisitor>::visit_TranslationUnit(LCompilers::ASR::TranslationUnit_t const&)
    this->visit_symbol(*a.second);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 5137, in LCompilers::ASR::BaseVisitor<LCompilers::InitExprVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 4850, in ??
    case symbolType::Program: { v.visit_Program((const Program_t &)x); return; }
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/init_expr.cpp", line 155, in LCompilers::InitExprVisitor::visit_Program(LCompilers::ASR::Program_t const&)
    visit_SymbolsContainingSymbolTable();
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/init_expr.cpp", line 147, in LCompilers::InitExprVisitor::visit_SymbolsContainingSymbolTable()
    this->visit_symbol(*sym);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 5137, in LCompilers::ASR::BaseVisitor<LCompilers::InitExprVisitor>::visit_symbol(LCompilers::ASR::symbol_t const&)
    void visit_symbol(const symbol_t &b) { visit_symbol_t(b, self()); }
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 4859, in ??
    case symbolType::Variable: { v.visit_Variable((const Variable_t &)x); return; }
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/init_expr.cpp", line 202, in LCompilers::InitExprVisitor::visit_Variable(LCompilers::ASR::Variable_t const&)
    call_replacer();
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/init_expr.cpp", line 121, in LCompilers::InitExprVisitor::call_replacer()
    replacer.replace_expr(*current_expr);
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/asr.h", line 46077, in LCompilers::ASR::BaseExprReplacer<LCompilers::ReplaceInitExpr>::replace_expr(LCompilers::ASR::expr_t*)
    self().replace_ArrayConstant(down_cast<ASR::ArrayConstant_t>(x));
  File "/home/harshita/Desktop/lfortran/src/libasr/pass/init_expr.cpp", line 49, in LCompilers::ReplaceInitExpr::replace_ArrayConstant(LCompilers::ASR::ArrayConstant_t*)
    PassUtils::ReplacerUtils::replace_ArrayConstant(x, this,
  File "/home/harshita/Desktop/lfortran/src/libasr/../libasr/pass/pass_utils.h", line 855, in ??
    LCOMPILERS_ASSERT_MSG(ASRUtils::extract_n_dims_from_ttype(result_var_type) == 1,
  File "./stdlib/abort.c", line 79, in __GI_abort()
  File "./signal/../sysdeps/posix/raise.c", line 26, in __GI_raise()
  File "./nptl/pthread_kill.c", line 43, in __pthread_kill_implementation()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __GI___sigaction()
Abort: Signal SIGABRT (abort) received

Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #4574.

### CODE REVIEW

**Review by HarshitaKalani — APPROVED** [review]
Looks perfect. Thank you!

**Comment by Pranavchiku** [comment]
The tests work well for me on mac. I'll check if again.

File: src/libasr/pass/array_op.cpp [hunk 1]
```diff
@@ -1728,8 +1728,11 @@ class ArrayOpVisitor : public ASR::CallReplacerOnExpressionsVisitor<ArrayOpVisit
             ASR::expr_t** current_expr_copy_9 = current_expr;
             ASR::expr_t* original_value = x.m_value;
             if( ASR::is_a<ASR::ArrayBroadcast_t>(*x.m_value) ) {
-                resultvar2value[replacer.result_var] =
-                    ASR::down_cast<ASR::ArrayBroadcast_t>(original_value)->m_array;
+                ASR::ArrayBroadcast_t* array_broadcast = ASR::down_cast<ASR::ArrayBroadcast_t>(x.m_value);
+                if ( ASR::is_a<ASR::TypeInquiry_t>(*array_broadcast->m_array) ) {
+                    array_broadcast->m_array = ASR::down_cast<ASR::TypeInquiry_t>(array_broadcast->m_array)->m_value;
+                }
+                resultvar2value[replacer.result_var] = array_broadcast->m_array;
```
Comment (Pranavchiku) [inline src/libasr/pass/array_op.cpp:1735]: This ensures the broadcast to be similar to `array = <scalar-constant>`

**Review by certik — APPROVED** [review]

**Comment by HarshitaKalani** [comment]
The PR is good. We can merge this after resolving the conflicts.

**Pull request merged**